### PR TITLE
Added dual-phase case in the cosmic ray vertex building algorithm

### PR DESCRIPTION
**Pull request by echardon**

### CODE REVIEW

File: larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc [hunk 1]
```diff
@@ -172,43 +173,77 @@ void CosmicRayVertexBuildingAlgorithm::BuildCosmicRayParent(const LArPointingClu
                 maxDirection = pointingCluster.GetOuterVertex().GetDirection();
             }
             else
-            {
-                LArClusterHelper::GetExtremalCoordinates(pCluster, minPosition, maxPosition);
+	    {
+	        LArClusterHelper::GetExtremalCoordinates(pCluster, minPosition, maxPosition);
```
Comment (johnmarshall80) [inline larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc]: (Cosmetic) Whitespace-only change.

File: larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc [hunk 2]
```diff
@@ -172,43 +173,77 @@ void CosmicRayVertexBuildingAlgorithm::BuildCosmicRayParent(const LArPointingClu
...
                 minDirection = (maxPosition - minPosition).GetUnitVector();
                 maxDirection = (minPosition - maxPosition).GetUnitVector();
-            }
-
+	    }
+
```
Comment (johnmarshall80) [inline larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc]: (Cosmetic) Whitespace-only change.

File: larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc [hunk 3]
```diff
@@ -172,43 +173,77 @@ void CosmicRayVertexBuildingAlgorithm::BuildCosmicRayParent(const LArPointingClu
...
-
-            if (!foundVtx || (minPosition.GetY() > std::max(maxPosition.GetY(), vtxPosition.GetY())))
-            {
-                foundVtx = true;
-                vtxPosition = minPosition;
-                vtxDirection = minDirection;
-            }
+   	        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+	    
+	    if (m_isDualPhase) //X is the vertical direction in Dual-Phase geometry
```
Comment (johnmarshall80) [inline larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc]: (Cosmetic) Indentation slightly different to that in surrounding code.

File: larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc [hunk 4]
```diff
@@ -172,43 +173,77 @@ void CosmicRayVertexBuildingAlgorithm::BuildCosmicRayParent(const LArPointingClu
...
+	    {
+		if (!foundVtx || (minPosition.GetX() > std::max(maxPosition.GetX(), vtxPosition.GetX())))
```
Comment (johnmarshall80) [inline larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayVertexBuildingAlgorithm.cc]: This implementation looks sound to me. The only change I wondered about (and feel free to to dismiss this) might be to cache the actual coordinates used in the comparisons upfront, storing y values by default and x values if in dual phase mode. The logic itself then doesn't need to be repeated, and the choice of using y or x as inputs to the main logic is clarified a little, upfront.

e.g.

```
const float minCoordinate(m_isDualPhase ? minPosition.GetX() : minPosition.GetY());
const float maxCoordinate(m_isDualPhase ? maxPosition.GetX() : maxPosition.GetY());

if (minCoordinate < maxCoordinate) // or some other, similar logic
    outputChosenVector = minPosition;
```

**Comment by PandoraPFA** [comment]
Hi Etienne,

Many thanks! I just started a new branch, based off the tip of the master branch (which corresponds to v03_15_12, tagged by larsoft), and cherry-picked your commits into this branch. I also made a few whitespace changes and updated the version (to v03_15_13) and changelog. I just pushed this to the branch ``feature/larpandoracontent_v03_15_13``. Can you, at your convenience, take a look at this branch and check/test that it still does everything you need? Thanks!

If this branch is all good, it's a matter of testing for e.g. ProtoDUNE-SP and ProtoDUNE-DP in larsoft (no change for SP, expected changes for DP), ensuring that there's a dunetpc branch with any modified XML files in it, then e-mailing the scisoft team to ask that they merge this branch.

Previously we'd push the branch to Redmine and they would just merge it from there. Now, for the first time, we'll need to find out how the new system works (with larsoft on GitHub). We'll have to work this out together, and there were some presentations discussing the new setup in this week's larsoft coordination meeting: https://indico.fnal.gov/event/22931/

(We might need to fork larsoft's larpandoracontent repository, merge the commits from our clone here into the fork and submit a PR to larsoft, but all the intellectual work and discussion should already have been done here. Previously was just a quick push from here to Redmine and an e-mail. It might look more fiddly now, but will essentially be the same process.)

Thanks again!

**Comment by PandoraPFA** [comment]
Hi, Just an update. I had productive discussions with some of the scisoft team yesterday, and the LArSoft repository for this package (github.com/LArSoft/larpandoracontent) is now a fork of this repository (github.com/PandoraPFA/LArContent). Once you're happy with the feature branch "feature/larpandoracontent_v03_15_13" in your testing, you can just let us know and we can just issue a simple pull request to the LArSoft repository and it's all done. Nice and simple!

If you want to test the branch "feature/larpandoracontent_v03_15_13" in your existing larsoft installation, please note that you can switch to the larpandoracontent directory and issue the following commands, or similar:

`
git remote add github https://github.com/PandoraPFA/LArContent.git
git fetch github
git checkout github/feature/larpandoracontent_v03_15_13 -b mylocalbranchname
`

If you then see any issues when running e.g. mrbsetenv, you can make local changes to the package versions listed in your larpandoracontent/ups/product_deps file, but we don't (commit or) push these, as they are managed by the scisoft team.

Hope this helps!

**Comment by echardon** [comment]
Hi ! I did my testing everything works fine on my side! 

On the other hand, I have the sample of 10k muons ready. Unfortunately things are a little bit more complicated for cosmics and neutrinos. Concerning cosmics I am confident it will be ready for next Monday. I am talking with Alex Himmel about putting some files on StashCache. For neutrinos things are a more blurry, I think all I need is to finish the prestaging but until I haven't take a look at the files themselves I can't be sure if it will be ready for monday. Should be ready during the week anyway.

Thank you very much!

**Comment by PandoraPFA** [comment]
Hi, thanks very much for testing the current feature branch. Great that it works for ProtoDUNE-DP. Does everything seem unchanged with e.g. ProtoDUNE-SP too?

Do you have a feature branch for dunetpc too, with any XML-related changes in there? You could add DP-specific details to pandoramodules_dune.fcl if it would help you. For instance, you could define e.g. a `protodune_pandora_dp` producer? Once branches are ready for larpandoracontent and dunetpc, then it's over to LArSoft and dune.

XML files for dune go [here](https://cdcvs.fnal.gov/redmine/projects/dunetpc/repository/revisions/develop/show/dune/DUNEPandora/scripts)
Pandora fcl configurations for dune defined [here](https://cdcvs.fnal.gov/redmine/projects/dunetpc/repository/revisions/develop/entry/dune/DUNEPandora/pandoramodules_dune.fcl)

We typically create a feature branch on dunetpc Redmine repo with the same name as the larpandoracontent feature, e.g. `feature/larpandoracontent_v03_15_13` in this case.

I'd suggest that we submit this to LArSoft and dune for the week after next. Next week is in principle fine for us, but it's the first week with a LArSoft release from GitHub and waiting one week should help keep things simple for the scisoft team. Does this sound OK?

Thanks for the info about the ProtoDUNE-DP sample production too!

**Comment by PandoraPFA** [comment]
Merged via larsoft repository.

**Pull request closed without merging**